src/timing.py

```diff
@@ -0,0 +1,370 @@
+    Returns:
+        List of elapsed times in milliseconds
+    """
+    if device is None:
+        if verbose:
+            print(f"Using current device: {torch.cuda.current_device()}")
+        device = torch.cuda.current_device()
+
+    # Warm ups
+    for _ in range(num_warmup):
+        kernel_fn(*args)
+        torch.cuda.synchronize(device=device)
+
+    # note this only release PyTorch’s CUDA caching allocator, not necessarily clearing device's L2 cache
+    torch.cuda.empty_cache()
+
+    print(f"[Profiling] Using device: {device} {torch.cuda.get_device_name(device)}, warm up {num_warmup}, trials {num_trials}"
+    )
+
+    elapsed_times: list[float] = [] # in ms
+
+    # Timing trials
+    for trial in range(num_trials + discard_first):
+        torch.cuda.synchronize(device=device) # block on all streams
+
+        # create event marker default is not interprocess
+        start_event = torch.cuda.Event(enable_timing=True)
+        end_event = torch.cuda.Event(enable_timing=True)
+        
+        clear_l2_cache() # measuring cold cache performance
+        
+        # note cuda events mark event on current stream
+        start_event.record()
+        _ = kernel_fn(*args)
+        end_event.record() 
+
+        # waits for all streams on that device
+        # though it is important to note the events only record time between on current stream
+        # TODO: find ways to check hacks by launching work on additional stream
+        torch.cuda.synchronize(device=device)
+
+        # Calculate the elapsed time in milliseconds
+        elapsed_time_ms = start_event.elapsed_time(end_event)
+        if trial >= discard_first:
+            if verbose:
+                logical_idx = trial - discard_first + 1
+                print(f"Trial {logical_idx}: {elapsed_time_ms:.3g} ms")
+            elapsed_times.append(elapsed_time_ms)
+
+    return elapsed_times
+
+
+def time_execution_with_do_bench_interface(
+    kernel_fn: callable,
+    args: list[Any],
+    # this is different for triton do_bench
+    num_warmup: int = 3, 
+    num_trials: int = 10,
+    discard_first: int = 1, # not used yet
+    verbose: bool = True,
+    device: torch.device | None = None) -> list[float]:
+    """
+    Using triton's default do_bench as it is 
+    Note we don't set num_warmup and num_trials, and we use warmup 25 ms and repetition time 100 ms with Triton's default values
+    See doc: https://triton-lang.org/main/python-api/generated/triton.testing.do_bench.html
+    Benchmark the runtime of the provided function. By default, return the median runtime of fn along with the 20-th and 80-th performance percentile.
+    """
+    do_bench_fn = lambda : kernel_fn(*args) # wrap function with arguments
+    return triton_testing.do_bench(fn=do_bench_fn,
+            warmup=25,
+            rep=100, 
+            grad_to_none=None, 
+            quantiles=None, 
+            return_mode="all")
+
+
+def time_execution_with_do_bench_impl(
```

Comment:
I think we can actually just remove the do_bench implementation (unless we want to add amd or smth). The usecase for it now is the extra backends, otherwise it is functionally the same as the time with cuda event